src/core/lib.lua

```diff
@@ -686,6 +686,38 @@ function getenv (name)
    else return nil end
 end
 
+function fstruct(def)
+   local struct = {}
+   local offset = 0
+   for ct, fld in def:gmatch('([%a_][%w_]*)%s+([%a_][%w_]*);') do
+      ct = ffi.typeof(ct)
+      struct[fld] = {
+         fieldname = fld,
+         ct = ct,
+         size = ffi.sizeof(ct),
+         offset = offset,
+      }
+      offset = offset + struct[fld].size
+   end
+   return struct, offset
+end
+
+function fieldrd(field, fd)
+   local buf = ffi.typeof('$ [1]', field.ct)()
+   local r, err = fd:pread(buf, field.size, field.offset)
+   if not r then error(err) end
+   return buf[0]
+end
+
+function fieldwr(field, fd, val)
+   local buf = ffi.typeof('$ [1]', field.ct)()
```

Comment:
The virtio PCI BAR0 access is done on initialization. At runtime, only the device notification is done. And with this approach it will be slow.
Unfortunately, the mmap of rersource0 for virtio-pci is not failing. It is not entirely clear why,though. 

By inlining - do you mean to just replace `ffi.typeof('$ [1]', field.ct)` with `ffi.typeof('struct pci_whatever')`?